cabal-testsuite/src/Test/Cabal/TestCode.hs

```diff
@@ -19,22 +27,52 @@ import Data.Typeable     (Typeable)
 data TestCode
     = TestCodeOk
     | TestCodeSkip String
-    | TestCodeKnownFail Int
-    | TestCodeUnexpectedOk
+    | TestCodeKnownFail IssueID
+    | TestCodeUnexpectedOk IssueID
     | TestCodeFail
+    | TestCodeFlakyFailed IssueID
+    | TestCodeFlakyPassed IssueID
   deriving (Eq, Show, Read, Typeable)
 
 instance Exception TestCode
   where
     displayException = displayTestCode
 
 displayTestCode :: TestCode -> String
-displayTestCode TestCodeOk            = "OK"
-displayTestCode (TestCodeSkip msg)    = "SKIP " ++ msg
-displayTestCode (TestCodeKnownFail t) = "OK (known failure, see #" <> show t <> ")"
-displayTestCode TestCodeUnexpectedOk  = "FAIL (unexpected success)"
-displayTestCode TestCodeFail          = "FAIL"
+displayTestCode TestCodeOk               = "OK"
+displayTestCode (TestCodeSkip msg)       = "SKIP " ++ msg
+displayTestCode (TestCodeKnownFail t)    = "OK (known failure, see " <> show t <> ")"
+displayTestCode (TestCodeUnexpectedOk t) = "FAIL (unexpected success, see " <> show t <> ")"
+displayTestCode TestCodeFail             = "FAIL"
+displayTestCode (TestCodeFlakyFailed t)  = "FLAKY (FAIL, see " <> show t <> ")"
+displayTestCode (TestCodeFlakyPassed t)  = "FLAKY (OK, see " <> show t <> ")"
 
 isTestCodeSkip :: TestCode -> Bool
 isTestCodeSkip (TestCodeSkip _) = True
 isTestCodeSkip _                = False
+
+type TestPassed = Bool
+
+newtype IssueID = IssueID Int
+  deriving newtype (Eq, Typeable, Num)
+
+instance Show IssueID where
+  show (IssueID i) = "#" <> show i
+
+instance Read IssueID where
```

Comment:
Apart from adhering to the roundtripping law for `read.show`, is there a reason to define this? If not, as I said above, you could probably avoid `Show` and, therefore, avoid this instance.

Reply:
The test code output is read by the testing ghci servers to determine whether a test passes or not. Unless I defined this it would say that it can't parse some outcomes.

But I think show could be the stock show and I could pretty show with some other function.

Reply:
Right… Yeah, of there was a way to avoid touching Show/Read, I'd prefer that. It's probably less code too. But if not, that's not the end of the world of course.

Reply:
Removed